src/main/java/com/mongodb/kafka/connect/source/schema/BsonDocumentToSchema.java

```diff
@@ -111,6 +137,134 @@ private static Schema inferSchema(final String fieldPath, final BsonValue bsonVa
+      return null;
+    }
+
+    SchemaBuilder builder = SchemaBuilder.struct().name(firstSchema.name()).optional();
+
+    for (Field firstField : firstSchema.fields()) {
+      Field secondField = secondSchema.field(firstField.name());
+      if (secondField == null || firstField.schema().equals(secondField.schema())) {
+        builder.field(firstField.name(), firstField.schema());
+      } else if (isSentinelValueSet(firstField.schema())) {
+        builder.field(secondField.name(), secondField.schema());
+      } else if (isSentinelValueSet(secondField.schema())) {
+        builder.field(firstField.name(), firstField.schema());
+      } else if (firstField.schema().type() == Schema.Type.STRUCT
+          && secondField.schema().type() == Schema.Type.STRUCT) {
+        Schema combinedSchema = combine(firstField.schema(), secondField.schema());
+        if (combinedSchema == null) {
+          LOGGER.debug(
+              "Can't combine non-matching struct fields: {} and {}", firstField, secondField);
+          return null;
+        }
+        builder.field(firstField.name(), combinedSchema);
+      } else if (firstField.schema().type() == Schema.Type.ARRAY
+          && secondField.schema().type() == Schema.Type.ARRAY) {
+        if (isSentinelValueSet(secondField.schema().valueSchema())) {
+          builder.field(firstField.name(), firstField.schema());
+        } else if (isSentinelValueSet(firstField.schema().valueSchema())) {
+          builder.field(secondField.name(), secondField.schema());
+        } else {
+          Schema combinedSchema =
+              combine(firstField.schema().valueSchema(), secondField.schema().valueSchema());
+          if (combinedSchema == null) {
+            LOGGER.debug(
+                "Can't combine non-matching array element value schema: {} and {}",
+                firstField,
+                secondField);
+            return null;
+          }
+          builder.field(
+              firstField.name(),
+              SchemaBuilder.array(combinedSchema).name(firstField.name()).optional().build());
+        }
+      } else {
+        LOGGER.debug("Can't combine non-matching fields: {} and {}", firstField, secondField);
+        return null;
+      }
+    }
+
+    for (Field secondField : secondSchema.fields()) {
+      if (firstSchema.field(secondField.name()) == null) {
+        builder.field(secondField.name(), secondField.schema());
+      }
+    }
+
+    return builder.build();
+  }
+
+  // Relies on the sentinel value added above
+  private static boolean isSentinelValueSet(final Schema schema) {
+    return schema.type() == Schema.Type.STRING
+        && SENTINEL_FOR_NULL_OR_EMPTY.equals(schema.defaultValue());
+  }
+
+  /**
+   * This method normalizes the schema that had been denormalized due to the array element
+   * schema-combining logic applied as part of schema building process.
+   */
+  private static Schema normalizeSchema(final Schema schema) {
```

Comment:
This shouldn't be needed as fields are already sorted (see createSchemaBuilder) so this looks to be duplicating work.